monitoring/mock_ridsp/routes_ridsp.py

```diff
@@ -1,30 +1,88 @@
+import arrow
 import datetime
+from typing import List, Optional
 
 import flask
+import s2sphere
 
-from monitoring.monitorlib import rid
+from monitoring.monitorlib import geo, rid
+from monitoring.monitorlib.rid_automated_testing.injection_api import TestFlight
+from monitoring.monitorlib.typing import StringBasedDateTime
 from monitoring.mock_ridsp import webapp
 from monitoring.mock_ridsp.auth import requires_scope
+from .database import db
 
 
-@webapp.route('/v1/uss/identification_service_areas/<id>', methods=['POST'])
+def _get_report(flight: TestFlight, t_request: datetime.datetime,
+                view: s2sphere.LatLngRect, include_recent_positions: bool) -> Optional[rid.RIDFlight]:
+  details = flight.get_details(t_request)
+  if not details:
+    return None
+
+  recent_states = flight.select_relevant_states(
+    view, t_request - rid.NetMaxNearRealTimeDataPeriod, t_request)
+  if not recent_states:
+    # No recent telemetry applicable to view
+    return None
+
+  recent_states.sort(key=lambda p: p.timestamp)
+  result = rid.RIDFlight(
+    id=details.id,
+    aircraft_type='NotDeclared', #TODO: Include aircraft_type in TestFlight API
+    current_state=recent_states[-1],
+    simulated=True)
+  if include_recent_positions:
+    recent_positions: List[rid.RIDRecentAircraftPosition] = []
+    for recent_state in recent_states:
+      recent_positions.append(rid.RIDRecentAircraftPosition(
+        time=recent_state.timestamp,
+        position=recent_state.position))
+    result.recent_positions = recent_positions
+  return result
+
+
+@webapp.route('/ridsp/v1/uss/identification_service_areas/<id>', methods=['POST'])
 @requires_scope([rid.SCOPE_WRITE])
 def notify_isa(id: str):
-  return 'Notification acknowledged', 204
+  return flask.jsonify({
+    'message': 'mock_ridsp never solicits subscription notifications'
+  }), 400
 
 
-@webapp.route('/v1/uss/flights', methods=['GET'])
+@webapp.route('/ridsp/v1/uss/flights', methods=['GET'])
 @requires_scope([rid.SCOPE_READ])
 def flights():
-  return flask.jsonify({
-    'timestamp': datetime.datetime.utcnow().isoformat() + 'Z',
-    'flights': []
-  })
+  if 'view' not in flask.request.args:
+    return flask.jsonify(rid.ErrorResponse(message='Missing required "view" parameter')), 400
+  try:
+    view = geo.make_latlng_rect(flask.request.args['view'])
+  except ValueError as e:
+    return flask.jsonify(rid.ErrorResponse(message='Error parsing view: {}'.format(e))), 400
+
+  include_recent_positions = flask.request.args.get('include_recent_positions', 'False').lower() == 'true'
+
+  diagonal = view.lo().get_distance(view.hi()).degrees * geo.EARTH_CIRCUMFERENCE_KM / 360
+  if diagonal > rid.NetMaxDisplayAreaDiagonal:
+    msg = 'Requested diagonal of {} km exceeds limit of {} km'.format(diagonal, rid.NetMaxDisplayAreaDiagonal)
+    return flask.jsonify(rid.ErrorResponse(message=msg)), 413
 
+  now = arrow.utcnow().datetime
+  flights = []
+  for test_id, record in db.tests.items():
```

Comment:
What happens if by mistake two tests are submitted (with same flight tracks) for the same view, this will give the data twice. I guess I am asking in this API, there is no check here against a test_id

Reply:
When its `flights` endpoint is queried, any RID Service Provider should respond with all flights it manages within the area of interest.  There is no concept of a "test_id" when querying an SP (using the ASTM-standard interface) for its flights, nor should there be -- the data source of flights known to the SP is internal to that SP.  If two tests with the same tracks are injected into a single SP, that SP should indeed respond with two flights at identical positions when its `flights` are queried.

We fundamentally want to create as few parallel code paths as possible that are exercised only during the test -- instead, we're trying to test the entire system end-to-end.  Since a production Display Provider will not provide a test_id when querying a Service Provider, we should not provide that test_id when testing either.

mock_ridsp should not assign the same flight ID to two different flights, however, and that is what would happen currently.  In a future PR, we'll add the ability to adjust the injected test data to fix problems like this (or perhaps to reject the injection attempt).